Python/Product/Analysis/Interpreter/Ast/AstPythonInterpreterFactory.cs

```diff
@@ -643,32 +680,67 @@ TryImportModuleContext context
 
             var mp = mmp.Value;
 
-#if USE_TYPESHED
-            lock (_typeShedPathsLock) {
-                if (_typeShedPaths == null) {
-                    var typeshed = FindModuleInSearchPath(_factory.GetSearchPaths(), _factory.GetImportableModules(), "typeshed");
-                    if (typeshed.HasValue) {
-                        _typeShedPaths = GetTypeShedPaths(PathUtils.GetParent(typeshed.Value.SourceFile)).ToArray();
-                    } else {
-                        _typeShedPaths = Array.Empty<string>();
-                    }
-                }
-                if (_typeShedPaths.Any()) {
-                    var mtsp = FindModuleInSearchPath(_typeShedPaths, null, mp.FullName);
-                    if (mtsp.HasValue) {
-                        mp = mtsp.Value;
-                    }
-                }
-            }
-#endif
+            IPythonModule module;
 
             if (mp.IsCompiled) {
                 _log?.Log(TraceLevel.Verbose, "ImportScraped", mp.FullName, FastRelativePath(mp.SourceFile));
-                return new AstScrapedPythonModule(mp.FullName, mp.SourceFile);
+                module = new AstScrapedPythonModule(mp.FullName, mp.SourceFile);
+            } else {
+                _log?.Log(TraceLevel.Verbose, "Import", mp.FullName, FastRelativePath(mp.SourceFile));
+                module = PythonModuleLoader.FromFile(context.Interpreter, mp.SourceFile, LanguageVersion, mp.FullName);
+            }
+
+            return module;
+        }
+
+        private IReadOnlyList<string> GetTypeShedPaths(int timeout) {
+            // First check
+            var typeShedPaths = _typeShedPaths;
+            if (typeShedPaths != null) {
+                return typeShedPaths;
+            }
+            lock (_searchPathsLock) {
+                // Second check after locking
+                if (_typeShedPaths != null) {
+                    return _typeShedPaths;
+                }
+            }
+
+            try {
+                var importTask = GetTypeShedPathsAsync();
+                return importTask.Wait(timeout) ? importTask.Result : null;
+            } catch (Exception ex) {
+                _log?.Log(TraceLevel.Error, "GetTypeShedPaths", ex.ToString());
+            }
+            return null;
+        }
+
+        private async Task<IReadOnlyList<string>> GetTypeShedPathsAsync() {
+            // Assume caller has already done the quick checks
+
+            IReadOnlyList<string> typeShedPaths;
+            var typeshed = await FindModuleInSearchPathAsync("typeshed");
```

Comment:
Is it ok that `FindModuleInSearchPathAsync` can be called several times concurrently before `_typeShedPaths` is finally assigned?

Reply:
Yep. We could call it every time if we wanted, it doesn't change any state. But since it's results (rarely) change, better to cache it.